software/firmware/europi.py

```diff
@@ -1,249 +1,278 @@
-        except:
-            raise Exception("EuroPi Software Error:\ncentre.text() only accepts string") #Display a useful error message
-        maximum_lines = (self.height / 9) // 1 #Lines are 9 pixels tall, and the //1 means it will calculate only the whole number of available lines
-        if len(lines) > maximum_lines:
-            raise Exception("EuroPi Software Error:\nDisplay is not big enough for that many lines") #Display a useful error message
-        else:
-            padding_top = (self.height - (len(lines) * 9)) / 2
-            for index in range(len(lines)):
-                content = lines[index]
-                padding_left = int((self.width - ((len(content)+1) * 7)) / 2)
-                self.text(content, padding_left-1, int((index * 9) + padding_top)-1)
-            oled.show()
-
-
-
-
-class Output: #Class used to control the CV outputs
-    def __init__(self, pin): #Takes only the GPIO pin being used for the output
-        self.output = PWM(Pin(pin))
-        self.output.freq(1000000) #This is specified as the default is too low to prevent audible PWM 'hum'
-        self.pin = pin
-        self.current_duty = 0
-        self.output_multiplier = get_output_calibration_data() #Uses the calibration data stored in calibration.txt to scale the duty into a voltage
-        
-    def current_voltage(self): #A method used to see what the current voltage of the output is
-        return self.current_duty / self.output_multiplier
-        
-    def duty(self, cycle): #A method to set the output based on a duty cycle of the PWM. Not very useful from a user perspective but required to be able to define voltage()
-        cycle = int(cycle)
-        self.output.duty_u16(clamp(cycle, 0, 65534)) #Clamps the duty cycle to 0-65534. By default values over or under that range will cycle over to the next limit which is not very useful
-        self.current_duty = cycle
-        
-    def voltage(self, voltage): #Set the output voltage based on a float voltage
-        self.duty(voltage * self.output_multiplier)
-        
-    def on(self): #Set the output voltage to 5. Useful if using the output as a faux-digital output
-        self.voltage(5)
-    
-    def off(self): #Turns off the output (0V) 
-        self.duty(0)
-    
-    def toggle(self): #Toggles the output between 0V and 5V depending on the current voltage.
-        if self.current_duty > 500: #Arbitrary duty cycle to compare against, but some value is required otherwise toggle would only work when it begins at exactly 0V or 5V
-            self.off()
-        else:
-            self.on()
-        
-    def value(self, value): #Sets the output to 0V or 5V depending on a binary input of 0 or 1. Useful if replicating the digital input for example
-        if value == 1:
-            self.on()
-        else:
-            self.off()
+# Default font is 8x8 pixel monospaced font.
+CHAR_WIDTH = 8
+CHAR_HEIGHT = 8
 
 
+# Helper functions.
 
+def clamp(value, low, high):

```

Comment:
Bring back `clamp()`'s documentation as a docstring

"Returns a value that is no lower than 'low' and no higher than 'high'"

Reply:
Done.